server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobStore.java

```diff
@@ -100,16 +102,25 @@ public void close() {
         Preconditions.checkNotNull(data);
         ObjectStorageBucketName resolvedBucketName = bucketNameResolver.resolve(bucketName);
 
-        BlobId blobId = blobIdFactory.forPayload(data);
-        Payload payload = payloadCodec.write(data);
-
-        Blob blob = blobStore.blobBuilder(blobId.asString())
-            .payload(payload.getPayload())
-            .contentLength(payload.getLength().orElse(Long.valueOf(data.length)))
-            .build();
+        return Mono.fromCallable(() -> blobIdFactory.forPayload(data))
+            .flatMap(blobId -> blobExistenceTester.exists(resolvedBucketName, blobId)
+                .flatMap(exists -> {
+                    if (exists) {
+                        return Mono.just(blobId);
+                    }
+                    return save(resolvedBucketName, blobId, data);
+                }));
+    }
 
-        return Mono.fromRunnable(() -> blobPutter.putDirectly(resolvedBucketName, blob))
-            .thenReturn(blobId);
+    private Mono<BlobId> save(ObjectStorageBucketName resolvedBucketName, BlobId blobId, byte[] data) {
+        return Mono.fromCallable(() -> payloadCodec.write(data))
```

Comment:
I think that depends on the payload codec.
DefaultPayloadCodec is not cpu intensive and might not deserve a context switch but AES one definitly do justify it.

Maybe having PayloadCodec reactive could be the way to go so that each payload codec can plan which scheduler to run on? (in a separate PR?)